doc/developer/design/20230210_storage_persist_sink_redux.md

```diff
@@ -0,0 +1,195 @@
+               _.-´ |       .-´            |      \
+           _.-´     |    .-´               |       \
+        .-´  .------+----|-------+---------|--------\-----.
+       /    /            |       |         |         \     \
+,--------------.   ,-----------------.     |     ,-----------------.
+| write_batches|   |  write_batches  |     |     |  write_batches  |
+| worker 0     |   | worker 1        |     |     | worker N        |
++-----+--------+   +-+---------------+     |     +--+--------------+
+       \              \                    |        /
+        `-.            `,                  |       /
+           `-._          `-.               |      /
+               `-._         `-.            |     /
+                   `---------. `-.         |    /
+                             +`---`---+-------------,
+                             | append_batches       |
+                             | one arbitrary worker |
+                             +------+---------------+
+```
+
+The different operators work like this:
+
+1. `mint_batch_descriptions` emits new batch descriptions whenever the frontier of source `Collection` advances.
+A batch description is a pair of `(previous_upper, upper)` that tells write operators which updates to write and
+in the end tells the append operator what frontiers to use for the `compare_and_append`. This is a single-worker operator.
+2. `write_batches` writes the source `Collection` to persist. It does this by creating a persist `BatchBuilder` for each
+timestamp, and writing every value that occurs at each timestamp to the corresponding `BatchBuilder`. This uses a bounded
+amount of memory. This does not yet append the batches to the persist shard, the update are only uploaded/prepared to be
+appended to a shard. We only emit finalized batches to the `append_batches` operator when that batch is contained within
+a batch descriptions that we learned about from `mint_batch_descriptions`.
+3. `append_batches` takes as input the minted batch descriptions and written batches. Whenever the frontiers sufficiently advance,
+we take a batch description and all the batches that belong to it and append it to the persist shard.
+This also updates the shared frontier information in `storage_state` to match the persist frontier.
+This way, the controller eventually learns of the upper frontier of data written to the persist shard.
+Note that failing to `compare_and_append` must be retried, as we current expect only 1 replica to be
+running this sink operation for each source `Collection`.
+
+### Notes on the implementation
+
+- It is needlessly expensive to create a `BatchBuilder` for each timestamp.
+  - A change to this implementation is to avoid spilling data to persist for timestamps
+  whose values use a small amount of memory. These can be consolidated into large batches in `append_batches`,
+  similar to how <https://github.com/MaterializeInc/materialize/pull/17526> works.
```

Comment:
Convince me that this doesn't introduce a correctness issue. What happens when we read data, but not write it to persist immediately and keep it in memory for a while? What happens when we crash? The answer might be it doesn't matter, but an explanation would be great!

Reply:
The batches we write but aren't yet `compare_and_append`'d are just dropped if we crash, and we rewrite new batches on restart, because we restart from the shard's upper (so, the last successful `compare_and_append`)